README.md
# IDOLclient [![Build Status](https://travis-ci.org/ColorfullyMe/idol-client.svg)](https://travis-ci.org/ColorfullyMe/idol-client) [![NPM](https://nodei.co/npm/idol-client.png)](https://nodei.co/npm/idol-client/)

Javascript client for working with the HP's IDOL OnDemand API from Node.js/io.js and the browser. Supports all endpoints with theirs parameters, can auto do polling of async tasks, fires events and more...

Made and maintained by <a href="https://github.com/ColorfullyMe" target="_blank">@Colorfully</a> for free. It is open source as software should one day be!:)

**Note** that this work is not oficial HP project!
# Installation
Available as npm package.

`[sudo] npm install idol-client`

Then in **Node.js:**

- `var MyClient = require('idol-client')(options)`

Or in the **browser:**

Include the file `build/idol-client.min.js` to your project. It will always be the latest version. For older ones checkout to previous tags.

- `<script src="build/idol-client.min.js"></script>`

# Contributing
Pull requsts welcome!

Fork. Make changes. Post pull requests...

Uses `gulp` as build tool. Tasks defined:

- `browser-build`  Runs browserify, uglify and etc. Puts result in the `/build` folder.
- `docs`  Creates documentation and this README.
- `APIref`  Parses source via `markdox` to create the API reference.

**Tests** available. Made with `mocha` and `chai`. To execute them:

`[sudo] npm test`

# How to support the project?
> It takes time to do it but saves time using it...

<a href="//gittip.com/ColorfullyMe" target="_blank">Gittip</a>

Fork. Make changes. Post pull requests...



<!-- Start lib/browser-export.js -->

## IDOLclient

<!-- End lib/browser-export.js -->




<!-- Start lib/idol-client.js -->

# API
This is the `idol-client` API reference.

## Constructor

## IDOLclient(options)

Factory/Constructor creating class instances based on `options` parameter. This is what one gets when the lib is required in Node.js or embeded as script in the browser.

In case `options` is a **String** then it should be the IDOL API key to use.
If you need to alter the default options provide an **Object**. Posible `options` are:

- **String** **Required** `APIkey`  IDOL's API key to use
- **Object** `APIformat`  Used to build API URLs
 - `host`  Defaults to 'api.idolondemand.com'
 - `platform`  Defaults to '1'
- **Boolean** `autoResult`  Automatically wait for async job result and fire result events. Defaults to `true`.
- **Falsey|Int** `requestTimeout`  Reject request promises if not fulfilled or rejected after `requestTimeout`. Defaults to false.

Examples:
```js
 var MyClient = require('idol-client')('MyRegisteredAPIkey');
 // or
 var MyClient = require('idol-client')({
     APIkey: 'MyRegisteredAPIkey'
 });
 // or in the browser
 var MyClient = IDOLclient('MyRegisteredAPIkey');
 ...
```

### Params:

* **String|Object** *options* 

## Properties

Because the lib integrates and depands on `lodash`, `Q` and `superagent` those libraries are included into the browser build. Thus it is useful to export and make them available for you. This way your apps using this lib will have `lodash`, `Q` and `superagent` loaded and ready for use. This is very useful in the browser.

- `_` Thea famous `lodash` [lodash](https://www.npmjs.org/package/lodash)
- `Q` The famous `Q` [q](https://www.npmjs.org/package/q)
- `superagent` The famous `superagent` [superagent](https://github.com/visionmedia/superagent)
- `Methods` Array of method names which are shortcuts to IDOL API endpoins [APIs](https://www.idolondemand.com/developer/apis)

Examples:
```js
 // in the browser...
 var MyClient = IDOLclient('MyRegisteredAPIkey');
 var _ = MyClient._;
 var Q = MyClient.Q;

 _.isElement(document.body)  // => true
 Q.isPromise({}) // => false
 ...
```

## Events
IDOLclient's instance is an event aggregator based on node's [EventEmitter](http://nodejs.org/api/events.html#events_class_events_eventemitter). One can use it to fire custom events and/or use the predefined ones:

- `job result`  Fired when some async job is ready with result.
- `job result error`  Fired when some async job errored in waiting for result.

Examples:
```js
 // in the browser...
 var MyClient = IDOLclient('MyRegisteredAPIkey');

 MyClient.on('job result', function(job){
     console.log('Your async job is ready!', job);
 });
 ...
```
## Methods
The library currently covers all available IDOL API [endpoints](https://www.idolondemand.com/developer/apis) with theirs parameters. To do so it uses the low level method `request(options)` described bellow. For convinience shortcuts to ease the pain when working with the lib are provided which are defined in the `./lib/core/shrotcuts.js`. They use the `request` method internally. There are currently **52** shortcuts:

- recognizeSpeech(options)
- connectorStatus(options)
- createConnector(options)
- deleteConnector(options)
- retrieveConfig(options)
- startConnector(options)
- updateConnector(options)
- ...
- and so on.

Where `options` is a **Object** with:

- **String** `method` The request method. Defaults to `GET`.
- **String** `type`  The request type. Can be `sync` or `async`. Defaults to `sync`.
- **String** `version`  The API version. Defaults to `v1`.
- **Object** `parameters`  Requst query and/or post parameters. The `APIkey` will be included automatically.
- **Object** `files`  Collection of files to attach to request. Only for `POST` and `PUT`.
- **Object** `headers`  Custom headers to set for the requst.
- **HTMLFormElement** `formElm`  Form element to process and attach to request. **Only in the browser**. See [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData).

For example to get the status of one of your indexes you could do:
```js
 var MyClient = require('idol-client')('MyRegisteredAPIkey');
 //or in the browser...
 var MyClient = IDOLclient('MyRegisteredAPIkey')

 MyClient.indexStatus({
     parameters: {
         index: 'myIndexName'
     }
 }).then(
     function(res){
         console.log(res.code);    // => 200-299
         console.log(res.headers);    // => Response headers
         console.log(res.data);    // => Response data
     },
     function(error){ console.log('Ups, some error occured:', error); }
 );
```

More complex example. Lets POST some file for highlighting:
```js
 var MyClient = require('idol-client')('MyRegisteredAPIkey');
 //or in the browser...
 var MyClient = IDOLclient('MyRegisteredAPIkey')

 // In Node.js...
 MyClient.highlightText({
     method: 'POST',
     parameters: {
         file: 'myFileName',
         highlight_expression: 'Terms to highlight'
     },
     files: {
         myFileName: __dirname+'/some/path/to/file/fileToAttach.doc' // See IDOL supported files.
     }
 }).then(
     function(res){
         console.log(res.code);    // => 200-299
         console.log(res.headers);    // => Response headers
         console.log(res.data);    // => Response data
     },
     function(error){ console.log('Ups, some error occured:', error); }
 );

 // In browser...
 MyClient.highlightText({
     method: 'POST',
     parameters: {
         file: 'myFileName',
         highlight_expression: 'Terms to highlight'
     },
     files: {
         myFileName: document.getElementById('fileSelect').files[0]  // Should be `File` or `Blob`!
     }
 }).then(
     function(res){
         console.log(res.code);    // => 200-299
         console.log(res.headers);    // => Response headers
         console.log(res.data);    // => Response data
     },
     function(error){ console.log('Ups, some error occured:', error); }
 );
```

In the example above one should note the difference how files are attached depending on the environment. In general:

- **Node.js** - `files` object is key/value pairs of attachment name of the file/path of the file. File will be loaded automatically from the file system.
- **Browser** - `files` object is key/value pairs of attachment name of the file/File or Blob objects. If you provide `HTMLFormElement` reference under the `formElm` instead of specifing each file separately in `files` the form will be processed and inputs attached to the request.

Additionally there are also the listed methods below...

<!-- End lib/idol-client.js -->




<!-- Start lib/core/jobEndpoint.js -->

## jobStatus(jobID)

Retrieve status of a job by UUID.

### Params:

* **String** *jobID* UUID of the job

### Return:

* **Object** Q.promise

## jobResult(jobID)

Waits for job result by UUID.

### Params:

* **String** *jobID* UUID of the job

### Return:

* **Object** Q.promise

## jobAsync(actions, files)

Submit(POST) an asynchronous job to IDOL API for processing.

### Params:

* **Array** *actions* Array of `Action` objects to execute. [See here](https://www.idolondemand.com/developer/docs/AsynchronousAPI.htm)
* **Object** *files* Collection of files to attach to the POST request

### Return:

* **Object** Q.promise

<!-- End lib/core/jobEndpoint.js -->




<!-- Start lib/core/request.js -->

## request(options)

Low level method to make requests to the IDOL API.

### Params:

* **Object** *options* The request options

### Return:

* **Object** Q.promise

<!-- End lib/core/request.js -->




<!-- Start lib/core/shortcuts.js -->

<!-- End lib/core/shortcuts.js -->



docs/HISTORY.md
0.3.3 / 2015-05-30
==================

  * Deploy fixes.

0.3.3 / 2015-05-28
==================

  * Minor jshint fixes.

0.3.3 / 2015-05-21
==================

  * Travis add iojs for tests.
  * Fixes auto npm plublish issues with old npm version.

0.3.2 / 2015-04-15
==================

  * Version bump

0.3.1 / 2015-04-01
==================

  * New available API endpoints added.
  * JSHint config file.

0.2.1 / 2015-02-09
==================

  * New available API endpoints added.
  * Automatically release on NPM.

0.1.0 / 2014-12-01
==================

  * Initial commit


test/core.js
if (typeof window != 'object') {
    // Nodejs
    var CLIENT = require('../lib/idol-client')('someAPIkey'),
        FACTORY = require('../lib/idol-client'),
        assert = require('chai').assert,
        expect = require('chai').expect,
        Q = require('q');
} else {
    // Browser
    var CLIENT = IDOLclient('someAPIkey'),
        FACTORY = IDOLclient,
        assert = chai.assert,
        expect = chai.expect,
        Q = CLIENT.Q;
}

CLIENT.Q.longStackSupport = true;

/*
 * Test test test
 */
describe('IDOLclient core', function () {

    describe('IDOLclient externals', function () {
        it('should export lodash', function () {
            assert.isFunction(CLIENT._);
            assert.isTrue(CLIENT._.isPlainObject({}), 'should be lodash indeed');
        });

        it('should export Q', function () {
            assert.isFunction(CLIENT.Q);
            assert.isFalse(Q.isPromise(), 'should be Q indeed');
        });

        it('should export superagent', function () {
            assert.isFunction(CLIENT.superagent);
        });
    });

    describe('IDOLclient.request(options)', function () {
        it('should be method in the class', function () {
            assert.isFunction(CLIENT.request);
        });

        it('should throw without options', function () {
            assert.throw(CLIENT.request);
        });

        it('should throw when the APIdentifier is missing', function () {
            try {
                CLIENT.request({});
                assert.fail();
            } catch (e) {
                assert.instanceOf(e, Error);
                assert.equal(e.message, 'IDOLclient.request(options) need API name as option key `API`');
            }
        });

        it('should throw when the API request type is wrong', function () {
            try {
                CLIENT.request({
                    API: 'some API',
                    type: 'not sync or async'
                });
                assert.fail();
            } catch (e) {
                assert.instanceOf(e, Error);
                assert.equal(e.message, 'API request type should be `sync` or `async`');
            }
        });

        it('should always return promise. It will always be rejected on HTTPS connection errors', function (done) {
            var ins = FACTORY({
                APIkey: 'someAPIkey',
                APIformat: {
                    host: 'some-wrong-host.com',
                    platform: '1'
                }
            });
            ins.request({
                API: 'some API',
            }).done(
                assert.fail,
                function (e) {
                    assert.instanceOf(e, Error);
                    done();
                });
        });

        it('should always return promise. It will always be rejected when http response code <200 && >=300 with rsp like object', function (done) {
            var p = CLIENT.request({
                API: 'some API'
            });

            assert.isTrue(Q.isPromise(p));

            p.done(
                assert.fail,
                function (rsp) {
                    assert.isObject(rsp);
                    expect(rsp.code).to.satisfy(function (code) {
                        return code < 200 || code >= 300;
                    });
                    assert.isObject(rsp.headers);
                    assert.isObject(rsp.data);
                    done();
                }
            );
        });
    });
});


test/examples.js
// Those tests require IDOL API key to run with.
// Try to load it from .env file...
require('dotenv').load();
if (!process.env.IDOL_API_KEY)
    return; // No key no tests.

// Create a client instance to use.
var MyClient = require('../lib/idol-client.js')(process.env.IDOL_API_KEY),
    assert = require('chai').assert;

MyClient.Q.longStackSupport = true;


describe('Async request[s]', function () {

    it('resolve with the `jobID`, trigger `job result` event when job result available but only when `autoResult` is truthy', function (done) {
        // Send some async job
        // if `autoResult` is truthy events will fire
        // when result is available|errored.
        MyClient.highlightText({
            type: 'async',
            parameters: {
                text: 'some test text to highlight',
                highlight_expression: 'test'
            }
        }).done(
            function (r) {
                assert.isObject(r);
                assert.isObject(r.data);
                assert.isString(r.data.jobID);
                // If `autoResult` is falsy no `job result` events are fired
                // thus exit here.
                if (!MyClient.options.autoResult)
                    done();
                //console.log(r);
            },
            assert.fail
        );

        MyClient.on('job result', function (res) {
            //console.log(res);
            assert.isObject(res);
            assert.isObject(res.headers);
            assert.isNumber(res.code);
            assert.property(res.data, 'status');
            assert.property(res.data, 'jobID');
            assert.property(res.data, 'actions');
            assert.isString(res.data.status);
            assert.isString(res.data.jobID);
            assert.isArray(res.data.actions);
            done();
        });

        MyClient.on('job result error', assert.fail);
    });

    it('can be processed via Q easily too', function (done) {
        // Send some async job
        // process it via promise resolution and chaining.
        var checkID;
        MyClient.jobAsync([{
            name: 'highlight',
            version: 'v1',
            params: {
                file: 'myfile',
                highlight_expression: 'test'
            }
        }], {
            myfile: __dirname + '/test.doc'
        })
            .get('data')
            .get('jobID')
            .tap(function (theID) {
                checkID = theID;
            })
            .then(MyClient.jobResult)
            .get('data')
            .done(function (jobresult) {
                assert.isString(jobresult.status);
                assert.equal(jobresult.status, 'finished');
                assert.isString(jobresult.jobID);
                assert.isArray(jobresult.actions);
                assert.equal(jobresult.jobID, checkID);
                done();
            }, assert.fail);
    });
});

describe('Sync request[s]', function () {
    it('resolve with the result in the promise', function (done) {
        MyClient.highlightText({
            parameters: {
                text: 'some test text to highlight',
                highlight_expression: 'test'
            }
        })
            .get('data')
            .get('text')
            .done(
                function (text) {
                    assert.equal(text, 'some <span style="background-color: yellow">test</span> text to highlight');
                    done();
                },
                assert.fail
        );
    });
});


test/shortcuts.js
if (typeof window != 'object') {
    // Nodejs
    var CLIENT = require('../lib/idol-client')('someAPIkey'),
        shortcuts = require('../lib/core/shortcuts'),
        assert = require('chai').assert,
        Q = require('q');
} else {
    // Browser
    var CLIENT = IDOLclient('someAPIkey'),
        shortcuts = CLIENT.Methods,
        assert = chai.assert,
        Q = CLIENT.Q;
}

CLIENT.Q.longStackSupport = true;

/*
 * Test test test
 */
describe('IDOLclient shortcuts', function () {

    it('should not have duplicates', function () {
        function hasDuplicates(array) {
            var valuesSoFar = {};
            for (var i = 0; i < array.length; ++i) {
                var value = array[i];
                if (Object.prototype.hasOwnProperty.call(valuesSoFar, value)) {
                    return true;
                }
                valuesSoFar[value] = true;
            }
            return false;
        }

        assert.isFalse(hasDuplicates(shortcuts));
    });

    // No need to repeat dummy tests for all methods
    // as they are created automatically and use same pattern
    // so instead just test a rendom method from the collection.
    var name = CLIENT._.sample(shortcuts);

    describe('IDOLclient.' + name + '(options)', function () {
        it('should be method in the class', function () {
            assert.isFunction(CLIENT[name]);
        });

        it('should throw without options', function () {
            assert.throw(CLIENT[name], 'IDOLclient.' + name + ' requires {} as first argument');
        });

        it('should throw when the API request type is wrong', function () {
            try {
                CLIENT[name]({
                    type: 'not sync or async'
                });
                assert.fail();
            } catch (e) {
                assert.instanceOf(e, Error);
                assert.equal(e.message, 'API request type should be `sync` or `async`');
            }
        });

        it('should not throw when the options is empty {}', function () {
            try {
                CLIENT[name]({});
            } catch (e) {
                assert.fail();
            }
        });
    });

    describe('IDOLclient.jobStatus(jobID)', function () {
        it('should be method in the class', function () {
            assert.isFunction(CLIENT.jobStatus);
        });

        it('should throw without options', function () {
            assert.throw(CLIENT.jobStatus, 'IDOLclient.jobStatus(jobID) requires UUID string as 1st argument');
        });
    });

    describe('IDOLclient.jobResult(jobID)', function () {
        it('should be method in the class', function () {
            assert.isFunction(CLIENT.jobResult);
        });

        it('should throw without options', function () {
            assert.throw(CLIENT.jobResult, 'IDOLclient.jobResult(jobID) requires UUID string as 1st argument');
        });
    });

    describe('IDOLclient.jobAsync(actions, files)', function () {
        it('should be method in the class', function () {
            assert.isFunction(CLIENT.jobAsync);
        });

        it('should throw without options', function () {
            assert.throw(CLIENT.jobAsync, 'IDOLclient.jobAsync(actions, files) requires none empty array with `Actions` as 1st argument');
        });
    });
});


gulpfile.js
var gulp = require('gulp'),
    header = require('gulp-header'),
    markdox = require('gulp-markdox'),
    uglify = require('gulp-uglify'),
    browserify = require('browserify'),
    source = require('vinyl-source-stream'),
    buffer = require('vinyl-buffer'),
    sourcemaps = require('gulp-sourcemaps'),
    concat = require('gulp-concat'),
    pkg = require('./package.json'),
    banner = ['/**',
                ' * <%= pkg.name %>',
                ' * @version v<%= pkg.version %>',
                ' * @license <%= pkg.licenses[0].type %>',
                ' * @license Multiple others by Lodash, Q and superagent',
                ' */\n'
             ].join('\n');


gulp.task('APIref', function () {
    return gulp.src('lib/**/*.js')
        .pipe(markdox())
        .pipe(concat('API.md'))
        .pipe(gulp.dest('./docs'));
});


gulp.task('docs', ['APIref'], function () {
    return gulp.src(['docs/README.md', 'docs/API.md'])
        .pipe(concat('README.md'))
        .pipe(gulp.dest('./'));
});


gulp.task('browser-build', function () {

    var bundler = browserify({
        entries: ['./lib/browser-export.js'],
        debug: true
    });

    var bundle = function () {
        return bundler
            .bundle()
            .pipe(source(pkg.name + '.' + 'min.js'))
            .pipe(buffer())
            .pipe(sourcemaps.init({
                loadMaps: true
            }))
            // Add transformation tasks to the pipeline here.
            .pipe(uglify())
            .pipe(header(banner, {
                pkg: pkg
            }))
            .pipe(sourcemaps.write('./'))
            .pipe(gulp.dest('./build/'));
    };

    return bundle();
});


lib/core/shortcuts.js
/*!
 * Represents all shortcuts provided as methods by the class
 */

module.exports = [

    /*!
     * Audio-Video Analytics
     */
    'recognizeSpeech',

    /*!
     * Connectors
     */
    'cancelConnectorSchedule',
    'connectorHistory',
    'connectorStatus',
    'createConnector',
    'deleteConnector',
    'retrieveConfig',
    'startConnector',
    'stopConnector',
    'updateConnector',

    /*!
     * Format Conversion
     */
    'expandContainer',
    //'ocrDocument', added under Image Analysis
    'storeObject',
    'extractText',
    'viewDocument',

    /*!
     * Image Analysis
     */
    'recognizeBarCodes',
    'detectFaces',
    'recognizeImages',
    'ocrDocument',

    /*!
     * Policy
     */
    'createClassificationObjects',
    'createPolicyObjects',
    'deleteClassificationObjects',
    'deletePolicyObjects',
    'retrieveClassificationObjects',
    'retrievePolicyObjects',
    'updateClassificationObjects',
    'updatePolicyObjects',

    /*!
     * Prediction
     */
    'predict',
    'recommend',
    'trainPredictor',

    /*!
     * Search
     */
    'createQueryProfile',
    'deleteQueryProfile',
    'queryTextIndex',
    'findRelatedConcepts',
    'findSimilar',
    'getContent',
    'getParametricValues',
    'retrieveIndexFields',
    'updateQueryProfile',

    /*!
     * Text Analysis
     */
    'classifyDocument',
    'categorizeDocument',
    'extractEntities',
    'expandTerms',
    'extractConcepts',
    'highlightText',
    'identifyLanguage',
    'analyzeSentiment',
    'tokenizeText',

    /*!
     * Unstructured Text Indexing
     */
    'addToTextIndex',
    'createTextIndex',
    'deleteTextIndex',
    'deleteFromTextIndex',
    'indexStatus',
    'listIndexes',
    'listResources',
    'restoreTextIndex',

    /*!
     * User Management
     */
    'addRole',
    'addStore',
    'addUser',
    'assignRole',
    'authenticate',
    'deleteRole',
    'deleteStore',
    'deleteUser',
    'listRoles',
    'listStores',
    'listUserRoles',
    'listUsers',
    'unassignRole',
    'verify'
];


lib/core/jobEndpoint.js
'use strict';

/*!
 * Module dependencies
 */
var _ = require('lodash');


module.exports = function (IDOLclient) {

    /**
     * Retrieve status of a job by UUID.
     *
     * @param   {String} jobID UUID of the job
     * @returns {Object} Q.promise
     * @api public
     */
    IDOLclient.prototype.jobStatus = function (jobID) {
        // Validate `jobID`.
        if (!jobID || !_.isString(jobID))
            throw new Error('IDOLclient.jobStatus(jobID) requires UUID string as 1st argument');

        return this.request({
            path: '/' + this.options.APIformat.platform + '/job/status/' + jobID
        });
    };

    /**
     * Waits for job result by UUID.
     *
     * @param   {String} jobID UUID of the job
     * @returns {Object} Q.promise
     * @api public
     */
    IDOLclient.prototype.jobResult = function (jobID) {
        // Validate `jobID`.
        if (!jobID || !_.isString(jobID))
            throw new Error('IDOLclient.jobResult(jobID) requires UUID string as 1st argument');

        return this.request({
            path: '/' + this.options.APIformat.platform + '/job/result/' + jobID
        });
    };

    /**
     * Submit(POST) an asynchronous job to IDOL API for processing.
     *
     * @param   {Array} actions Array of `Action` objects to execute. [See here](https://www.idolondemand.com/developer/docs/AsynchronousAPI.htm)
     * @param   {Object} files Collection of files to attach to the POST request
     * @returns {Object} Q.promise
     * @api public
     */
    IDOLclient.prototype.jobAsync = function (actions, files) {
        // Validate `actions` and `files`.
        if (!_.isArray(actions) || _.isEmpty(actions))
            throw new Error('IDOLclient.jobAsync(actions, files) requires none empty array with `Actions` as 1st argument');
        if (files && !_.isPlainObject(files))
            throw new Error('IDOLclient.jobAsync(actions, files) requires {} as 2nd argument');

        return this.request({
            method: 'POST',
            path: '/' + this.options.APIformat.platform + '/job',
            parameters: {
                job: {
                    actions: actions
                }
            },
            files: files
        });
    };
};


lib/idol-client.js
'use strict';

/*!
 * Module dependencies
 */
var _ = require('lodash'),
    Q = require('q'),
    EventEmitter = require('events').EventEmitter;

module.exports = function createIDOLclient(options) {

    // Create new instance of the class.
    var ins = new IDOLclient(options);

    // Bind all methods to class context.
    _.bindAll(ins);

    // Export external libraries used as they are included
    // in the build anyway. So one can use them too.
    ins._ = _;
    ins.Q = Q;
    ins.superagent = require('superagent');
    // And shortcut names too.
    ins.Methods = require('./core/shortcuts');

    // Return the instance.
    return ins;
};

/**
 * # API
 * This is the `idol-client` API reference.
 *
 * ## Constructor
 */
/**
 * Factory/Constructor creating class instances based on `options` parameter. This is what one gets when the lib is required in Node.js or embeded as script in the browser.
 *
 * In case `options` is a **String** then it should be the IDOL API key to use.
 * If you need to alter the default options provide an **Object**. Posible `options` are:
 *
 * - **String** **Required** `APIkey`  IDOL's API key to use
 * - **Object** `APIformat`  Used to build API URLs
 *  - `host`  Defaults to 'api.idolondemand.com'
 *  - `platform`  Defaults to '1'
 * - **Boolean** `autoResult`  Automatically wait for async job result and fire result events. Defaults to `true`.
 * - **Falsey|Int** `requestTimeout`  Reject request promises if not fulfilled or rejected after `requestTimeout`. Defaults to false.
 *
 * Examples:
 *```js
 *  var MyClient = require('idol-client')('MyRegisteredAPIkey');
 *  // or
 *  var MyClient = require('idol-client')({
 *      APIkey: 'MyRegisteredAPIkey'
 *  });
 *  // or in the browser
 *  var MyClient = IDOLclient('MyRegisteredAPIkey');
 *  ...
 *```
 * @param {String|Object} options
 * @api public
 */
function IDOLclient(options) {

    // Validate default client options.
    if (options && _.isString(options)) {
        options = {
            APIkey: options
        };
    } else if (!_.isPlainObject(options)) {
        throw new Error('IDOLclient requires constructor options to be set as \'your APIkey\' or {}');
    } else if (!_.has(options, 'APIkey')) {
        throw new Error('IDOLclient needs `APIkey` property to be set');
    }

    // Handle default client options.
    this.options = _.defaults(options, {
        APIformat: {
            host: 'api.idolondemand.com',
            platform: '1'
        },
        autoResult: true,
        requestTimeout: false
    });

    // Add events functionality to the class.
    EventEmitter.call(this);
}
// Add events functionality to the class.
IDOLclient.prototype = Object.create(EventEmitter.prototype);

// Load all class core methods manualy
// as browserify will fail otherwise.
// Not sure!?
// TODO: test fs.readSync if works in the browser?
// probably not...
require('./core/request')(IDOLclient);
require('./core/jobEndpoint')(IDOLclient);


/**
 * ## Properties
 *
 * Because the lib integrates and depands on `lodash`, `Q` and `superagent` those libraries are included into the browser build. Thus it is useful to export and make them available for you. This way your apps using this lib will have `lodash`, `Q` and `superagent` loaded and ready for use. This is very useful in the browser.
 *
 * - `_` Thea famous `lodash` [lodash](https://www.npmjs.org/package/lodash)
 * - `Q` The famous `Q` [q](https://www.npmjs.org/package/q)
 * - `superagent` The famous `superagent` [superagent](https://github.com/visionmedia/superagent)
 * - `Methods` Array of method names which are shortcuts to IDOL API endpoins [APIs](https://www.idolondemand.com/developer/apis)
 *
 * Examples:
 *```js
 *  // in the browser...
 *  var MyClient = IDOLclient('MyRegisteredAPIkey');
 *  var _ = MyClient._;
 *  var Q = MyClient.Q;
 *
 *  _.isElement(document.body)  // => true
 *  Q.isPromise({}) // => false
 *  ...
 *```
 *
 * ## Events
 * IDOLclient's instance is an event aggregator based on node's [EventEmitter](http://nodejs.org/api/events.html#events_class_events_eventemitter). One can use it to fire custom events and/or use the predefined ones:
 *
 * - `job result`  Fired when some async job is ready with result.
 * - `job result error`  Fired when some async job errored in waiting for result.
 *
 * Examples:
 *```js
 *  // in the browser...
 *  var MyClient = IDOLclient('MyRegisteredAPIkey');
 *
 *  MyClient.on('job result', function(job){
 *      console.log('Your async job is ready!', job);
 *  });
 *  ...
 *```
 * ## Methods
 * The library currently covers all available IDOL API [endpoints](https://www.idolondemand.com/developer/apis) with theirs parameters. To do so it uses the low level method `request(options)` described bellow. For convinience shortcuts to ease the pain when working with the lib are provided which are defined in the `./lib/core/shrotcuts.js`. They use the `request` method internally. There are currently **52** shortcuts:
 *
 * - recognizeSpeech(options)
 * - connectorStatus(options)
 * - createConnector(options)
 * - deleteConnector(options)
 * - retrieveConfig(options)
 * - startConnector(options)
 * - updateConnector(options)
 * - ...
 * - and so on.
 *
 * Where `options` is a **Object** with:
 *
 * - **String** `method` The request method. Defaults to `GET`.
 * - **String** `type`  The request type. Can be `sync` or `async`. Defaults to `sync`.
 * - **String** `version`  The API version. Defaults to `v1`.
 * - **Object** `parameters`  Requst query and/or post parameters. The `APIkey` will be included automatically.
 * - **Object** `files`  Collection of files to attach to request. Only for `POST` and `PUT`.
 * - **Object** `headers`  Custom headers to set for the requst.
 * - **HTMLFormElement** `formElm`  Form element to process and attach to request. **Only in the browser**. See [FormData](https://developer.mozilla.org/en-US/docs/Web/API/FormData).
 *
 * For example to get the status of one of your indexes you could do:
 *```js
 *  var MyClient = require('idol-client')('MyRegisteredAPIkey');
 *  //or in the browser...
 *  var MyClient = IDOLclient('MyRegisteredAPIkey')
 *
 *  MyClient.indexStatus({
 *      parameters: {
 *          index: 'myIndexName'
 *      }
 *  }).then(
 *      function(res){
 *          console.log(res.code);    // => 200-299
 *          console.log(res.headers);    // => Response headers
 *          console.log(res.data);    // => Response data
 *      },
 *      function(error){ console.log('Ups, some error occured:', error); }
 *  );
 *```
 *
 * More complex example. Lets POST some file for highlighting:
 *```js
 *  var MyClient = require('idol-client')('MyRegisteredAPIkey');
 *  //or in the browser...
 *  var MyClient = IDOLclient('MyRegisteredAPIkey')
 *
 *  // In Node.js...
 *  MyClient.highlightText({
 *      method: 'POST',
 *      parameters: {
 *          file: 'myFileName',
 *          highlight_expression: 'Terms to highlight'
 *      },
 *      files: {
 *          myFileName: __dirname+'/some/path/to/file/fileToAttach.doc' // See IDOL supported files.
 *      }
 *  }).then(
 *      function(res){
 *          console.log(res.code);    // => 200-299
 *          console.log(res.headers);    // => Response headers
 *          console.log(res.data);    // => Response data
 *      },
 *      function(error){ console.log('Ups, some error occured:', error); }
 *  );
 *
 *  // In browser...
 *  MyClient.highlightText({
 *      method: 'POST',
 *      parameters: {
 *          file: 'myFileName',
 *          highlight_expression: 'Terms to highlight'
 *      },
 *      files: {
 *          myFileName: document.getElementById('fileSelect').files[0]  // Should be `File` or `Blob`!
 *      }
 *  }).then(
 *      function(res){
 *          console.log(res.code);    // => 200-299
 *          console.log(res.headers);    // => Response headers
 *          console.log(res.data);    // => Response data
 *      },
 *      function(error){ console.log('Ups, some error occured:', error); }
 *  );
 *```
 *
 * In the example above one should note the difference how files are attached depending on the environment. In general:
 *
 * - **Node.js** - `files` object is key/value pairs of attachment name of the file/path of the file. File will be loaded automatically from the file system.
 * - **Browser** - `files` object is key/value pairs of attachment name of the file/File or Blob objects. If you provide `HTMLFormElement` reference under the `formElm` instead of specifing each file separately in `files` the form will be processed and inputs attached to the request.
 *
 * Additionally there are also the listed methods below...
 */


// Add shortcut for each APIdentifier
require('./core/shortcuts').forEach(function (APIdentifier) {
    IDOLclient.prototype[APIdentifier] = function (options) {
        if (!_.isPlainObject(options)){
            throw new Error('IDOLclient.' + APIdentifier + ' requires {} as first argument');
        }
        return this.request(_.extend(options, {
            API: APIdentifier.toLowerCase()
        }));
    };
});
